File: Cpp/src/MainWindow.cpp
#include "MainWindow.h"
#include <QDragMoveEvent>
#include <QDebug>
#include <QPainter>
#include <QAction>
#include <QFileDialog>
#include <fstream>

MainWindow::MainWindow(QWidget *parent)
: QMainWindow(parent)
, m_currentPoint(0)
, m_timer(std::make_unique<QTimer>(this))
, m_percents(0)
, m_toolBar(std::make_unique<QToolBar>("Linker ToolBar"))
, m_drawAnimation(false)
, m_drawLines(false)
{
  FillToolBar();

  qsrand(time(0));
  setAcceptDrops(true);


  connect(m_timer.get(), SIGNAL(timeout()), this, SLOT(repaint()));
  m_timer->start(20);

  addToolBar(Qt::BottomToolBarArea, m_toolBar.get());
}

void MainWindow::AddPoint(const QPoint& coords)
{
  m_percents = 0;
  m_points.emplace_back(new MovablePoint(this, 10));
  m_points.back()->show();
  size_t idx = m_points.size() - 1;
  m_colors.emplace_back(qrand()%255, qrand()%255, qrand()%255);

  auto& point = m_points.back();

  connect(point.get(), &MovablePoint::moving, this, [idx, this]{ this->m_currentPoint = idx;});
  connect(point.get(), &MovablePoint::positionChanged, this, &MainWindow::CalcBezierCurve);

  point->move(coords.x() - point->width() / 2, coords.y() - point->height() / 2);

  update();
}

void MainWindow::mouseDoubleClickEvent(QMouseEvent *event)
{
  if (event->button() == Qt::LeftButton) {
    AddPoint(QPoint(event->pos().x(), event->pos().y()));
  }
}

void MainWindow::dragMoveEvent(QDragMoveEvent *event)
{
  auto& point = m_points[m_currentPoint];
  point->move(event->pos().x() - point->width() / 2, event->pos().y() - point->height() / 2);
  update();
  event->accept();
}


void MainWindow::dropEvent(QDropEvent *event)
{
  event->setDropAction(Qt::MoveAction);
  event->accept();
}

void MainWindow::dragEnterEvent(QDragEnterEvent *event)
{
  event->accept();
}

void MainWindow::dragLeaveEvent(QDragLeaveEvent *event)
{
  event->accept();
}

void MainWindow::paintEvent(QPaintEvent *event)
{
  if (m_points.empty())
    return;

  m_percents %= 100;

  Q_UNUSED(event);
  QPainter painter(this);

  painter.drawPixmap(rect(), m_bezierCurvPM);

  std::vector<QPointF> points;
  points.reserve(m_points.size());

  for (auto & point : m_points)
  {
    auto x = point->pos().x() + point->width() / 2;
    auto y = point->pos().y() + point->height() / 2;
    points.emplace_back(x,y);
  }

  double ratio = m_percents / 100.;

  for (unsigned j = 0; j < m_points.size() - 1; j++)
  {
    QLineF line(points[j],points[j + 1]);
    if (m_drawLines || m_drawAnimation)
      painter.drawLine(line);
    points[j] = line.pointAt(ratio);
  }

  if (!m_drawAnimation)
    return;

  for (unsigned i = 2; i < m_points.size(); ++i)
  {
    painter.setPen(QPen(m_colors[i], 2, Qt::SolidLine, Qt::RoundCap));
    for (unsigned j = 0; j < m_points.size() - i; j++)
    {
      QLineF line(points[j],points[j + 1]);
      painter.drawLine(line);
      points[j] = line.pointAt(ratio);
    }
    if (i == m_points.size() - 1)
    {
      painter.setPen(QPen(Qt::blue, 10, Qt::SolidLine, Qt::RoundCap));
      painter.drawPoint(points[0]);
    }
  }
  m_percents++;
}

void MainWindow::CalcBezierCurve()
{
  m_percents = 0;
  m_bezierCurvPM.fill(Qt::transparent);
  std::vector<QPointF> points;
  points.reserve(m_points.size());
  std::optional<QPointF> prevPoint;
  for (unsigned percent = 0; percent <= 100; percent++)
  {
    double ratio = percent / 100.;
    for (auto &point: m_points) {
      auto x = point->pos().x() + point->width() / 2;
      auto y = point->pos().y() + point->height() / 2;
      points.emplace_back(x, y);
    }
    for (unsigned i = 1; i < m_points.size(); ++i) {
      for (unsigned j = 0; j < m_points.size() - i; j++) {
        QLineF line(points[j], points[j + 1]);
        points[j] = line.pointAt(ratio);
      }
      if (i == m_points.size() - 1) {
        QPainter mapPainter(&m_bezierCurvPM);
        mapPainter.setPen(QPen(Qt::red, 5, Qt::SolidLine, Qt::RoundCap));
        if (prevPoint)
        {
          mapPainter.drawLine(prevPoint.value(), points[0]);
        }
        prevPoint = points[0];
      }
    }
  }
}

void MainWindow::resizeEvent(QResizeEvent *event)
{
  m_bezierCurvPM = QPixmap(size());
  m_bezierCurvPM.fill(Qt::transparent);
  CalcBezierCurve();

  QWidget::resizeEvent(event);
}

void MainWindow::FillToolBar()
{
  m_toolBar->addAction("Open file",[this]{LoadFromFile(QFileDialog::getOpenFileName(this,
                                                                                    tr("Load Curve"),
                                                                                    "",
                                                                                    tr("Bezier Curve Files (*.bzcv)")));
  } );

  auto * saveAction = new QAction("Save as");
  connect(saveAction, &QAction::triggered, this, [this]{
    QFileDialog dialog(this, "Save Curve", QString(),
                       "Bezier Curve Files (*.bzcv)");
    dialog.setDefaultSuffix(".bzcv");
    dialog.setAcceptMode(QFileDialog::AcceptSave);
    dialog.exec();
    SaveToFile(dialog.selectedFiles().front());
  });
  m_toolBar->addAction(saveAction);
  m_toolBar->addSeparator();

  auto * animationAction = new QAction("Animation");
  animationAction->setCheckable( true );
  connect(animationAction, &QAction::changed, this, [this]{
                                                                                           m_drawAnimation = !m_drawAnimation;
                                                                                           m_percents = 0;
                                                                                          });
  m_toolBar->addAction(animationAction);
  m_toolBar->addSeparator();

  auto * linesAction = new QAction("Connect dots");
  linesAction->setCheckable( true );
  connect(linesAction, &QAction::changed, this, [this]{ m_drawLines = !m_drawLines;});
  m_toolBar->addAction(linesAction);
  m_toolBar->addSeparator();

  m_toolBar->addAction("Clear", [this]{m_points.clear(); m_colors.clear();});
}

void MainWindow::LoadFromFile(const QString &fileName)
{
  m_points.clear();
  m_colors.clear();

  std::ifstream infile(fileName.toStdString());
  int x, y;
  while (infile >> x >> y)
  {
    AddPoint(QPoint(x, y));
  }
}

void MainWindow::SaveToFile(const QString &fileName)
{
  std::ofstream outFile(fileName.toStdString());
  for(auto& point : m_points)
  {
    outFile << point->x() << ' ' <<  point->y() << std::endl;
  }
}

File: Python/MainWindow.py
import os
import re
from random import randint

from PyQt5.QtWidgets import QMainWindow, QToolBar, QFileDialog, QAction
from PyQt5.QtCore import QTimer, Qt, QPoint, QPointF, QLineF
from moveble_point import MovablePoint
from PyQt5.QtGui import QColor, QPainter, QPen, QPixmap
from PyQt5 import QtGui


class BezierCurveWindow(QMainWindow):
    __current_point: int
    __timer: QTimer
    __percents: int
    __toolbar: QToolBar
    __draw_animation: bool
    __draw_lines: bool

    def __init__(self, parent=None):
        super().__init__(parent)
        self.__current_point = 0
        self.__timer = QTimer(self)
        self.__percents = 0
        self.__points = []
        self.__colors = []
        self.__bezier_curv_PM = QPixmap(self.size())
        self.__toolbar = QToolBar("Linker ToolBar")
        self.__draw_animation = False
        self.__draw_lines = False

        self.setAcceptDrops(True)
        self.fill_tool_bar()
        self.__timer.timeout.connect(self.repaint)
        self.__timer.start(20)
        self.addToolBar(Qt.BottomToolBarArea, self.__toolbar)

    def on_animation(self):
        self.__draw_animation = not self.__draw_animation
        self.__percents = 0

    def on_draw_lines(self):
        self.__draw_lines = not self.__draw_lines

    def on_clear(self):
        for point in self.__points:
            point.deleteLater()
        self.__points.clear()
        self.__colors.clear()
        self.repaint()

    def fill_tool_bar(self):
        self.__toolbar.addAction("Open file", lambda: self.load_from_file(QFileDialog.getOpenFileName(self, "Load "
                                                                                                            "Curve",
                                                                                                      "",
                                                                                                      "Bezier Curve "
                                                                                                      "Files ("
                                                                                                      "*.bzcv)")[0]))
        self.__toolbar.addSeparator()

        self.__toolbar.addAction("Save as", lambda: self.save(QFileDialog.getSaveFileName(self, "Save Curve",
                                                                                                "",
                                                                                                "Bezier Curve "
                                                                                                "Files ("
                                                                                                "*.bzcv)")[0]))
        self.__toolbar.addSeparator()

        self.__animation_action = QAction("Animation")
        self.__animation_action.setCheckable(True)
        self.__animation_action.changed.connect(self.on_animation)
        self.__toolbar.addAction(self.__animation_action)
        self.__toolbar.addSeparator()

        self.__draw_lines_action = QAction("Connect dots")
        self.__draw_lines_action.setCheckable(True)
        self.__draw_lines_action.changed.connect(self.on_draw_lines)
        self.__toolbar.addAction(self.__draw_lines_action)
        self.__toolbar.addSeparator()

        self.__clear_action = QAction("Clear")
        self.__clear_action.triggered.connect(self.on_clear)
        self.__toolbar.addAction(self.__clear_action)
        self.__toolbar.addSeparator()

    def add_point(self, coords: QPoint) -> None:
        self.__percents = 0
        self.__points.append(MovablePoint(self, 10))
        self.__colors.append(QColor(randint(0, 255), randint(0, 255), randint(0, 255)))
        idx = len(self.__points) - 1
        point: MovablePoint = self.__points[-1]
        point.show()
        point.moving.connect(lambda: self.set_current_point(idx))
        point.position_changed.connect(self.calc_bezier_curve)
        point.move(coords.x() - point.width() / 2, coords.y() - point.height() / 2)
        self.update()

    def set_current_point(self, idx: int) -> None:
        self.__current_point = idx

    def mouseDoubleClickEvent(self, event: QtGui.QMouseEvent) -> None:
        if event.button() == Qt.LeftButton:
            self.add_point(QPoint(event.pos().x(), event.pos().y()))

    def dragMoveEvent(self, event: QtGui.QDragMoveEvent) -> None:
        point: MovablePoint = self.__points[self.__current_point]
        point.move(event.pos().x() - point.width() / 2, event.pos().y() - point.height() / 2)
        self.update()
        event.accept()

    def dropEvent(self, event: QtGui.QDropEvent) -> None:
        event.setDropAction(Qt.MoveAction)
        event.accept()

    def dragEnterEvent(self, event: QtGui.QDragEnterEvent) -> None:
        event.accept()

    def dragLeaveEvent(self, event: QtGui.QDragLeaveEvent) -> None:
        event.accept()

    def paintEvent(self, event: QtGui.QPaintEvent) -> None:
        if len(self.__points) == 0:
            return
        self.__percents %= 100
        painter = QPainter(self)
        painter.drawPixmap(self.rect(), self.__bezier_curv_PM)
        points = []
        for point in self.__points:
            x = point.pos().x() + point.width() / 2
            y = point.pos().y() + point.height() / 2
            points.append(QPointF(x, y))
        ratio = self.__percents / 100
        for i in range(len(self.__points) - 1):
            line = QLineF(points[i], points[i + 1])
            if self.__draw_lines or self.__draw_animation:
                painter.drawLine(line)
            points[i] = line.pointAt(ratio)
        if not self.__draw_animation:
            return

        for i in range(2, len(self.__points)):
            painter.setPen(QPen(self.__colors[i], 2, Qt.SolidLine, Qt.RoundCap))
            for j in range(len(self.__points) - i):
                line = QLineF(points[j], points[j + 1])
                painter.drawLine(line)
                points[j] = line.pointAt(ratio)
            if i == len(self.__points) - 1:
                painter.setPen(QPen(Qt.blue, 10, Qt.SolidLine, Qt.RoundCap))
                painter.drawPoint(points[0])
        self.__percents += 1

    def resizeEvent(self, event: QtGui.QResizeEvent) -> None:
        self.calc_bezier_curve()
        super(BezierCurveWindow, self).resizeEvent(event)

    def calc_bezier_curve(self) -> None:
        self.__bezier_curv_PM = QPixmap(self.size())
        self.__bezier_curv_PM.fill(Qt.transparent)
        map_painter = QPainter(self.__bezier_curv_PM)
        map_painter.setPen(QPen(Qt.red, 5, Qt.SolidLine, Qt.RoundCap))

        self.__percents = 0
        points = []
        prev_point = None
        for percent in range(0, 101):
            ratio = percent / 100
            for point in self.__points:
                x = point.pos().x() + point.width() / 2
                y = point.pos().y() + point.height() / 2
                points.append(QPointF(x, y))
            for i in range(1, len(self.__points)):
                for j in range(len(self.__points) - i):
                    line = QLineF(points[j], points[j + 1])
                    points[j] = line.pointAt(ratio)
                if i == len(self.__points) - 1:
                    if prev_point:
                        map_painter.drawLine(prev_point, points[0])
                    prev_point = points[0]

    def load_from_file(self, file_name: str):
        if len(file_name) == 0:
            return
        self.on_clear()
        with open(file_name, "r") as f:
            for line in f:
                coord = [int(x) for x in line.split()]
                self.add_point(QPoint(coord[0], coord[1]))

    def save(self, file_name: str):
        if len(file_name) == 0:
            return
        _, extension = os.path.splitext(file_name)
        if len(extension) == 0:
            file_name += ".bzcv"

        with open(file_name, "w") as f:
            for point in self.__points:
                f.write(f"{point.x()} {point.y()}\n")